### ISSUE TITLE
Perl dies on usage of index [-1] for array in subroutine call

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#13072]
Migrated from [rt.perl.org#118691](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=118691) (status was 'resolved')

Searchable as RT118691$

**Comment by p5pRT** [comment]
### From wolf-dietrich_moeller@t-online.de

Hi\,
the following behaviour of Perl looks like a bug to me&#8203;:

In the test program below\, the third and fourth call to the subroutine
should behave equal\, i\.e\. the fourth call should not result in a crash of
Perl with the following message&#8203;:
"Modification of non\-creatable array value attempted\, subscript \-1 at \<PATH
TO PERL PROGRAM> line 17\."

Reason for my assumption is that the  " // undef" in the first call and the
third call to "func" should be superfluous\, as the non\-existing array
element should return "undef" anyhow\. Thus the third call and the fourth
call should behave identical\.

Test program&#8203;:
\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#
\# Test program to show crash of Perl in the fourth call to 'sub func'
\# Such crash seems to happen only when the index is negative and the array
element does not exist\.
\# For positive indices no crash\, even if the array element does not exist\.
use strict;
sub func \($\) \{ my $val = shift; print $val // 'undef'\," in sub func\\n" \}
my @&#8203;array;
my $x = $array\[0\];
my $y = $array\[\-1\];
print '@&#8203;array\[0\,\-1\] : '\,join\(' '\,map $\_ // 'undef'\,$x\,$y\)\," in main\\n";
print '\($array\[0\] // undef\) :'\,"\\n";
func \($array\[0\] // undef\);
print '\($array\[0\]\) :'\,"\\n";
func \($array\[0\]\);
print '\($array\[\-1\] // undef\) :'\,"\\n";
func \($array\[\-1\] // undef\);
print '\($array\[\-1\]\) :'\,"\\n";
func \($array\[\-1\]\);
\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#

The output of the test program is&#8203;:
@&#8203;array\[0\,\-1\] : undef undef in main
\($array\[0\] // undef\) :
undef in sub func
\($array\[0\]\) :
undef in sub func
\($array\[\-1\] // undef\) :
undef in sub func
\($array\[\-1\]\) :
Modification of non\-creatable array value attempted\, subscript \-1 at \<PATH
TO PERL PROGRAM> line 17\.

\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#

My system is Win XP\, Service pack 3\, and I currently use the Perl
distribution from ActiveState
"ActivePerl\-5\.16\.3\.1603\-MSWin32\-x86\-296746\.msi"\.
Searching for 'array' in the new and open bugs lists did not result in a
similar bug report\, thus I assume this was not reported before\.

Thanks
Wolf

PS\.
Sorry I did not use the recommended tools for version reporting\, as nowadays
I only have access to Windows environment with ActiveState Perl windows
binary\.
I know that I normally should report this bug to ActiveState\, but I have the
bad experience\, that ActiveState did not respond on a bug report at all
\(which was handled immediately by you\, when I sent it to you later\)\.
Thus I send this bug report directly to you\, as I assume that this bug is
not a problem of the Windows port\, but of the basic Perl as provided by
perl\.org\.



---
Dr\. Wolf\-Dietrich Moeller
D\-81739 München\, Germany
mailto&#8203;:wolf\-dietrich\_moeller@&#8203;t\-online\.de

**Comment by p5pRT** [comment]
### From @jkeenan

On Sat Jun 29 06&#8203;:54&#8203;:53 2013\, wolf\-dietrich\_moeller@&#8203;t\-online\.de wrote&#8203;:
> Hi\,
> the following behaviour of Perl looks like a bug to me&#8203;:
> 
> In the test program below\, the third and fourth call to the subroutine
> should behave equal\, i\.e\. the fourth call should not result in a crash of
> Perl with the following message&#8203;:
> "Modification of non\-creatable array value attempted\, subscript \-1 at
\<PATH
> TO PERL PROGRAM> line 17\."
> 
> Reason for my assumption is that the  " // undef" in the first call
and the
> third call to "func" should be superfluous\, as the non\-existing array
> element should return "undef" anyhow\. Thus the third call and the fourth
> call should behave identical\.
> 
> Test program&#8203;:
> \#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#
> \# Test program to show crash of Perl in the fourth call to 'sub func'
> \# Such crash seems to happen only when the index is negative and the array
> element does not exist\.
> \# For positive indices no crash\, even if the array element does not exist\.
> use strict;
> sub func \($\) \{ my $val = shift; print $val // 'undef'\," in sub func\\n" \}
> my @&#8203;array;
> my $x = $array\[0\];
> my $y = $array\[\-1\];
> print '@&#8203;array\[0\,\-1\] : '\,join\(' '\,map $\_ // 'undef'\,$x\,$y\)\," in main\\n";
> print '\($array\[0\] // undef\) :'\,"\\n";
> func \($array\[0\] // undef\);
> print '\($array\[0\]\) :'\,"\\n";
> func \($array\[0\]\);
> print '\($array\[\-1\] // undef\) :'\,"\\n";
> func \($array\[\-1\] // undef\);
> print '\($array\[\-1\]\) :'\,"\\n";
> func \($array\[\-1\]\);
> \#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#
> 
> The output of the test program is&#8203;:
> @&#8203;array\[0\,\-1\] : undef undef in main
> \($array\[0\] // undef\) :
> undef in sub func
> \($array\[0\]\) :
> undef in sub func
> \($array\[\-1\] // undef\) :
> undef in sub func
> \($array\[\-1\]\) :
> Modification of non\-creatable array value attempted\, subscript \-1 at \<PATH
> TO PERL PROGRAM> line 17\.
> 

I don't think this is a bug in Perl\.  In fact\, it's a feature
specifically documented in 'perldoc perldiag'"

\#\#\#\#\#
Modification of non\-creatable array value attempted\, %s

\(F\) You tried to make an array value spring into existence\, and the
subscript was probably negative\, even counting from end of the array
backwards\.
\#\#\#\#\#

To explain this\, I would like to strip your demonstration code of
superfluous aspects\.  Among other things\, there is absolutely no reason
for a prototype on your subroutine\.  Since you're working in Perl
5\.16\.3\, you have the 'say' function available to you\, and I'll use that
to reduce the number of keystrokes in the code\.

\#\#\#\#\#
use strict;
use warnings;
use feature 'say';

my @&#8203;array;
func \($array\[0\] // undef\);
func \($array\[0\]\);
func \($array\[\-1\] // undef\);
func \($array\[\-1\]\);

sub func \{ my $val = shift; say $val // 'undef'; \}
\#\#\#\#\#

Perl's '//' operator tests the defined\-ness of the left\-hand side of the
operator\, returns that value if it is defined\, but returns the
right\-hand side if the left\-hand is not defined\.

In your third call to 'func\(\)'\, the '//' operator imposes a context on
the values to its left and to its right\.  In this context\, the left\-hand
value\, $array\[\-1\]\, is found to be not defined\, so whatever is on the
right\-hand side of '//' is returned\.  It so happens that what is on the
right\-hand side is the undefined value\.  So that is what is passed to
func\(\)\, and func\(\) handles it appropriately by printing the string
'undef' to STDOUT\.

Your fourth call is not equivalent to your third call\.  There's no '//'
operator imposing a context on $array\[\-1\]\.  Perl looks at that variable
directly and notes that it runs afoul of the error condition described
above\.  You're asking for $array\[\-1\] to spring to life\.  I'm sure that
there are others who can explain better than I can why we treat that as
a fatal error \-\- but fatal it is nonetheless\.

> PS\.
> Sorry I did not use the recommended tools for version reporting\, as
nowadays
> I only have access to Windows environment with ActiveState Perl windows
> binary\.

For reporting the version of Perl in which you are experiencing a
problem\, all you need to do is to include the output of 'perl \-V' in
your email to 'perlbug@&#8203;perl\.org'\.

Thank you very much\.
Jim Keenan

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @cpansprout

On Sat Jun 29 08&#8203;:06&#8203;:12 2013\, jkeenan wrote&#8203;:
> Your fourth call is not equivalent to your third call\.  There's no '//'
> operator imposing a context on $array\[\-1\]\.  Perl looks at that variable
> directly and notes that it runs afoul of the error condition described
> above\.  You're asking for $array\[\-1\] to spring to life\.  I'm sure that
> there are others who can explain better than I can why we treat that as
> a fatal error \-\- but fatal it is nonetheless\.

Whether this is a bug or no\, it is certainly surprising\-\-and fixable\.

It’s just a matter of tweaking the defer logic in pp\_aelem\.  Then rvalue
uses of $\_\[0\] within the function will just work\, but assignment to
$\_\[0\] will croak with the same error \(i\.e\.\, the error will be delayed as
long as possible\)\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From @jkeenan

On Sat Jun 29 10&#8203;:13&#8203;:54 2013\, sprout wrote&#8203;:
> On Sat Jun 29 08&#8203;:06&#8203;:12 2013\, jkeenan wrote&#8203;:
> > Your fourth call is not equivalent to your third call\.  There's no '//'
> > operator imposing a context on $array\[\-1\]\.  Perl looks at that variable
> > directly and notes that it runs afoul of the error condition described
> > above\.  You're asking for $array\[\-1\] to spring to life\.  I'm sure that
> > there are others who can explain better than I can why we treat that as
> > a fatal error \-\- but fatal it is nonetheless\.
> 
> Whether this is a bug or no\, it is certainly surprising\-\-and fixable\.
> 

I think what is more surprising is that the poster's second and fourth
cases behave differently\, i\.e\.\, that func\($array\[0\]\) has at least the
appearance of being well\-behaved\, whereas func\($array\[\-1\] generates a
fatal error\.

> It’s just a matter of tweaking the defer logic in pp\_aelem\.  Then rvalue
> uses of $\_\[0\] within the function will just work\, but assignment to
> $\_\[0\] will croak with the same error \(i\.e\.\, the error will be delayed as
> long as possible\)\.
> 

Do we really want to do that?

**Comment by p5pRT** [comment]
### From @cpansprout

On Sun Jun 30 05&#8203;:37&#8203;:19 2013\, jkeenan wrote&#8203;:
> On Sat Jun 29 10&#8203;:13&#8203;:54 2013\, sprout wrote&#8203;:
> > On Sat Jun 29 08&#8203;:06&#8203;:12 2013\, jkeenan wrote&#8203;:
> > > Your fourth call is not equivalent to your third call\.  There's no
'//'
> > > operator imposing a context on $array\[\-1\]\.  Perl looks at that
variable
> > > directly and notes that it runs afoul of the error condition described
> > > above\.  You're asking for $array\[\-1\] to spring to life\.  I'm sure that
> > > there are others who can explain better than I can why we treat
that as
> > > a fatal error \-\- but fatal it is nonetheless\.
> > 
> > Whether this is a bug or no\, it is certainly surprising\-\-and fixable\.
> > 
> 
> I think what is more surprising is that the poster's second and fourth
> cases behave differently\, i\.e\.\, that func\($array\[0\]\) has at least the
> appearance of being well\-behaved\,

That is a feature\, added by 68dc07451\.  One can pass nonexistent
elements to functions without having them vivified immediately\.  Only
when $\_\[0\] is modified inside the function is the element vivified\.

> whereas func\($array\[\-1\] generates a
> fatal error\.
> 
> > It’s just a matter of tweaking the defer logic in pp\_aelem\.  Then rvalue
> > uses of $\_\[0\] within the function will just work\, but assignment to
> > $\_\[0\] will croak with the same error \(i\.e\.\, the error will be delayed as
> > long as possible\)\.
> > 
> 
> Do we really want to do that?

I say Yes\! :\-\)

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From @khwilliamson

On 06/30/2013 08&#8203;:03 AM\, Father Chrysostomos via RT wrote&#8203;:
> On Sun Jun 30 05&#8203;:37&#8203;:19 2013\, jkeenan wrote&#8203;:
>> On Sat Jun 29 10&#8203;:13&#8203;:54 2013\, sprout wrote&#8203;:
>>> On Sat Jun 29 08&#8203;:06&#8203;:12 2013\, jkeenan wrote&#8203;:
>>>> Your fourth call is not equivalent to your third call\.  There's no
> '//'
>>>> operator imposing a context on $array\[\-1\]\.  Perl looks at that
> variable
>>>> directly and notes that it runs afoul of the error condition described
>>>> above\.  You're asking for $array\[\-1\] to spring to life\.  I'm sure that
>>>> there are others who can explain better than I can why we treat
> that as
>>>> a fatal error \-\- but fatal it is nonetheless\.
>>>
>>> Whether this is a bug or no\, it is certainly surprising\-\-and fixable\.
>>>
>>
>> I think what is more surprising is that the poster's second and fourth
>> cases behave differently\, i\.e\.\, that func\($array\[0\]\) has at least the
>> appearance of being well\-behaved\,
>
> That is a feature\, added by 68dc07451\.  One can pass nonexistent
> elements to functions without having them vivified immediately\.  Only
> when $\_\[0\] is modified inside the function is the element vivified\.
>
>> whereas func\($array\[\-1\] generates a
>> fatal error\.
>>
>>> It’s just a matter of tweaking the defer logic in pp\_aelem\.  Then rvalue
>>> uses of $\_\[0\] within the function will just work\, but assignment to
>>> $\_\[0\] will croak with the same error \(i\.e\.\, the error will be delayed as
>>> long as possible\)\.
>>>
>>
>> Do we really want to do that?
>
> I say Yes\! :\-\)
>

\+1

**Comment by p5pRT** [comment]
### From @rjbs

\* Father Chrysostomos via RT \<perlbug\-followup@&#8203;perl\.org> \[2013\-06\-30T10&#8203;:03&#8203;:06\]
> > > It’s just a matter of tweaking the defer logic in pp\_aelem\.  Then rvalue
> > > uses of $\_\[0\] within the function will just work\, but assignment to
> > > $\_\[0\] will croak with the same error \(i\.e\.\, the error will be delayed as
> > > long as possible\)\.
> > > 
> > 
> > Do we really want to do that?
> 
> I say Yes\! :\-\)

I think so\.  \(I am speaking here only as a user\.  So\, probably&#8203;: apply unless
somebody has a great objection\.\.\. which probably should've happened by now?\)

\-\- 
rjbs

**Comment by p5pRT** [comment]
### From @cpansprout

On Sat Jul 06 12&#8203;:39&#8203;:26 2013\, perl\.p5p@&#8203;rjbs\.manxome\.org wrote&#8203;:
> I think so\.  \(I am speaking here only as a user\.  So\, probably&#8203;: apply
> unless
> somebody has a great objection\.\.\. which probably should've happened by
> now?\)

Erm\, apply what?  Nobody has written a patch yet\. :\-\)

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From @rjbs

\* Father Chrysostomos via RT \<perlbug\-followup@&#8203;perl\.org> \[2013\-07\-06T16&#8203;:22&#8203;:08\]
> On Sat Jul 06 12&#8203;:39&#8203;:26 2013\, perl\.p5p@&#8203;rjbs\.manxome\.org wrote&#8203;:
> > I think so\.  \(I am speaking here only as a user\.  So\, probably&#8203;: apply
> > unless
> > somebody has a great objection\.\.\. which probably should've happened by
> > now?\)
> 
> Erm\, apply what?  Nobody has written a patch yet\. :\-\)

Patches when written\, of course\.

\-\- 
rjbs

**Comment by p5pRT** [comment]
### From @demerphq

On 6 July 2013 21&#8203;:38\, Ricardo Signes \<perl\.p5p@&#8203;rjbs\.manxome\.org> wrote&#8203;:
> \* Father Chrysostomos via RT \<perlbug\-followup@&#8203;perl\.org> \[2013\-06\-30T10&#8203;:03&#8203;:06\]
>> > > It’s just a matter of tweaking the defer logic in pp\_aelem\.  Then rvalue
>> > > uses of $\_\[0\] within the function will just work\, but assignment to
>> > > $\_\[0\] will croak with the same error \(i\.e\.\, the error will be delayed as
>> > > long as possible\)\.
>> > >
>> >
>> > Do we really want to do that?
>>
>> I say Yes\! :\-\)
>
> I think so\.  \(I am speaking here only as a user\.  So\, probably&#8203;: apply unless
> somebody has a great objection\.\.\. which probably should've happened by now?\)

If it means that $array\[\-1\] will not die in when used as an rvalue on
an empty @&#8203;array then personally I am in favor\.

cheers\,
Yves


\-\-
perl \-Mre=debug \-e "/just|another|perl|hacker/"

**Comment by p5pRT** [comment]
### From @cpansprout

On Sat Jul 06 13&#8203;:36&#8203;:42 2013\, perl\.p5p@&#8203;rjbs\.manxome\.org wrote&#8203;:
> \* Father Chrysostomos via RT \<perlbug\-followup@&#8203;perl\.org> \[2013\-07\-
> 06T16&#8203;:22&#8203;:08\]
> > On Sat Jul 06 12&#8203;:39&#8203;:26 2013\, perl\.p5p@&#8203;rjbs\.manxome\.org wrote&#8203;:
> > > I think so\.  \(I am speaking here only as a user\.  So\, probably&#8203;:
> apply
> > > unless
> > > somebody has a great objection\.\.\. which probably should've
> happened by
> > > now?\)
> >
> > Erm\, apply what?  Nobody has written a patch yet\. :\-\)
> 
> Patches when written\, of course\.

Speaking of patches\, LvTARGOFF is a STRLEN\, which is unsigned\, but I
need to store a signed value in there\.

Before I go and screw things up by casting between types of different
sizes\, what is the signed equivalent of STRLEN?  Is it SSize\_t?  Or
should I make xlv\_targoff into a STRLEN/SSize\_t union?

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From @cpansprout

On Tue Jul 16 18&#8203;:15&#8203;:32 2013\, sprout wrote&#8203;:
> On Sat Jul 06 13&#8203;:36&#8203;:42 2013\, perl\.p5p@&#8203;rjbs\.manxome\.org wrote&#8203;:
> > \* Father Chrysostomos via RT \<perlbug\-followup@&#8203;perl\.org> \[2013\-07\-
> > 06T16&#8203;:22&#8203;:08\]
> > > On Sat Jul 06 12&#8203;:39&#8203;:26 2013\, perl\.p5p@&#8203;rjbs\.manxome\.org wrote&#8203;:
> > > > I think so\.  \(I am speaking here only as a user\.  So\, probably&#8203;:
> > apply
> > > > unless
> > > > somebody has a great objection\.\.\. which probably should've
> > happened by
> > > > now?\)
> > >
> > > Erm\, apply what?  Nobody has written a patch yet\. :\-\)
> > 
> > Patches when written\, of course\.
> 
> Speaking of patches\, LvTARGOFF is a STRLEN\, which is unsigned\, but I
> need to store a signed value in there\.
> 
> Before I go and screw things up by casting between types of different
> sizes\, what is the signed equivalent of STRLEN?  Is it SSize\_t?  Or
> should I make xlv\_targoff into a STRLEN/SSize\_t union?

Attached is a patch to fix this bug\.  It also be found on the
sprout/negdefelem118691 branch\.  I’m not sure that what I’ve done is the
cleanest fix\.  I think \(SSize\_t\)LvTARGOFF would work\, resulting in a
smaller patch\.  Existing code uses \(I32\)\.

On both big\-endian and little\-endian machines I have tried casting \-1 to
unsigned long long and then to long\, which produces \-1\, so I may be
worrying about nothing\.  I’m no expert on these casting issues\, though\.

This patch happens to trigger bug \#7508\, causing a test failure\, so I
have added a dependency to the ticket\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From @cpansprout

On Tue Jul 16 23&#8203;:04&#8203;:12 2013\, sprout wrote&#8203;:
> On Tue Jul 16 18&#8203;:15&#8203;:32 2013\, sprout wrote&#8203;:
> > On Sat Jul 06 13&#8203;:36&#8203;:42 2013\, perl\.p5p@&#8203;rjbs\.manxome\.org wrote&#8203;:
> > > \* Father Chrysostomos via RT \<perlbug\-followup@&#8203;perl\.org> \[2013\-07\-
> > > 06T16&#8203;:22&#8203;:08\]
> > > > On Sat Jul 06 12&#8203;:39&#8203;:26 2013\, perl\.p5p@&#8203;rjbs\.manxome\.org wrote&#8203;:
> > > > > I think so\.  \(I am speaking here only as a user\.  So\, probably&#8203;:
> > > apply
> > > > > unless
> > > > > somebody has a great objection\.\.\. which probably should've
> > > happened by
> > > > > now?\)
> > > >
> > > > Erm\, apply what?  Nobody has written a patch yet\. :\-\)
> > > 
> > > Patches when written\, of course\.
> > 
> > Speaking of patches\, LvTARGOFF is a STRLEN\, which is unsigned\, but I
> > need to store a signed value in there\.
> > 
> > Before I go and screw things up by casting between types of different
> > sizes\, what is the signed equivalent of STRLEN?  Is it SSize\_t?  Or
> > should I make xlv\_targoff into a STRLEN/SSize\_t union?
> 
> Attached is a patch to fix this bug\.  It also be found on the
> sprout/negdefelem118691 branch\.  I’m not sure that what I’ve done is the
> cleanest fix\.  I think \(SSize\_t\)LvTARGOFF would work\, resulting in a
> smaller patch\.  Existing code uses \(I32\)\.
> 
> On both big\-endian and little\-endian machines I have tried casting \-1 to
> unsigned long long and then to long\, which produces \-1\, so I may be
> worrying about nothing\.  I’m no expert on these casting issues\, though\.
> 
> This patch happens to trigger bug \#7508\, causing a test failure\, so I
> have added a dependency to the ticket\.

And here is the patch\.


\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From @cpansprout

From 3361c5f90cc05b1e8baef2e8eccd3686ce27f70d Mon Sep 17 00&#8203;:00&#8203;:00 2001
From&#8203;: Father Chrysostomos \<sprout@&#8203;cpan\.org>
Date&#8203;: Tue\, 16 Jul 2013 22&#8203;:56&#8203;:44 \-0700
Subject&#8203;: \[PATCH\]

\[perl \#118691\] with targoff union

commit message not written yet

Not sure that the union is the best way to do it\.  It seems
\(SSize\_t\)LvTARGOFF and \(I32\)SvTARGOFF both work\, which would not require
as many changes\.

This triggers bug \#7508 in Peek\.t\.

<details><summary>Inline Patch</summary>

```diff
diff --git a/mg.c b/mg.c
index 99169cc..adf9b87 100644
--- a/mg.c
+++ b/mg.c
@@ -2316,10 +2316,10 @@ Perl_magic_getdefelem(pTHX_ SV *sv, MAGIC *mg)
             if (he)
                 targ = HeVAL(he);
 	}
-	else {
+	else if (LvSTARGOFF(sv) >= 0) {
 	    AV *const av = MUTABLE_AV(LvTARG(sv));
-	    if ((I32)LvTARGOFF(sv) <= AvFILL(av))
-		targ = AvARRAY(av)[LvTARGOFF(sv)];
+	    if (LvSTARGOFF(sv) <= AvFILL(av))
+		targ = AvARRAY(av)[LvSTARGOFF(sv)];
 	}
 	if (targ && (targ != &PL_sv_undef)) {
 	    /* somebody else defined it for us */
@@ -2370,14 +2370,16 @@ Perl_vivify_defelem(pTHX_ SV *sv)
 	if (!value || value == &PL_sv_undef)
 	    Perl_croak(aTHX_ PL_no_helem_sv, SVfARG(mg->mg_obj));
     }
+    else if (LvSTARGOFF(sv) < 0)
+	Perl_croak(aTHX_ PL_no_aelem, LvSTARGOFF(sv));
     else {
 	AV *const av = MUTABLE_AV(LvTARG(sv));
-	if ((I32)LvTARGLEN(sv) < 0 && (I32)LvTARGOFF(sv) > AvFILL(av))
+	if ((I32)LvTARGLEN(sv) < 0 && LvSTARGOFF(sv) > AvFILL(av))
 	    LvTARG(sv) = NULL;	/* array can't be extended */
 	else {
-	    SV* const * const svp = av_fetch(av, LvTARGOFF(sv), TRUE);
+	    SV* const * const svp = av_fetch(av, LvSTARGOFF(sv), TRUE);
 	    if (!svp || (value = *svp) == &PL_sv_undef)
-		Perl_croak(aTHX_ PL_no_aelem, (I32)LvTARGOFF(sv));
+		Perl_croak(aTHX_ PL_no_aelem, LvSTARGOFF(sv));
 	}
     }
     SvREFCNT_inc_simple_void(value);
diff --git a/pp_hot.c b/pp_hot.c
index 914a9d7..77f9805 100644
--- a/pp_hot.c
+++ b/pp_hot.c
@@ -2897,7 +2897,7 @@ PP(pp_aelem)
     IV elem = SvIV(elemsv);
     AV *const av = MUTABLE_AV(POPs);
     const U32 lval = PL_op->op_flags & OPf_MOD || LVRET;
-    const U32 defer = (PL_op->op_private & OPpLVAL_DEFER) && (elem > av_len(av));
+    const U32 defer = PL_op->op_private & OPpLVAL_DEFER;
     const bool localizing = PL_op->op_private & OPpLVAL_INTRO;
     bool preeminent = TRUE;
     SV *sv;
@@ -2938,14 +2938,20 @@ PP(pp_aelem)
 #endif
 	if (!svp || *svp == &PL_sv_undef) {
 	    SV* lv;
+	    IV len;
 	    if (!defer)
 		DIE(aTHX_ PL_no_aelem, elem);
+	    len = av_len(av);
 	    lv = sv_newmortal();
 	    sv_upgrade(lv, SVt_PVLV);
 	    LvTYPE(lv) = 'y';
 	    sv_magic(lv, NULL, PERL_MAGIC_defelem, NULL, 0);
 	    LvTARG(lv) = SvREFCNT_inc_simple(av);
-	    LvTARGOFF(lv) = elem;
+	    /* Resolve a negative index now, unless it points before the
+	       beginning of the array, in which case record it for error
+	       reporting in magic_setdefelem. */
+	    LvSTARGOFF(lv) =
+		elem < 0 && len + elem >= 0 ? len + elem : elem;
 	    LvTARGLEN(lv) = 1;
 	    PUSHs(lv);
 	    RETURN;
diff --git a/sv.h b/sv.h
index 1f9145f..782b43a 100644
--- a/sv.h
+++ b/sv.h
@@ -512,7 +512,10 @@ struct xpvlv {
     _XPV_HEAD;
     union _xivu xiv_u;
     union _xnvu xnv_u;
-    STRLEN	xlv_targoff;
+    union {
+	STRLEN	xlvu_targoff;
+	SSize_t xlvu_stargoff;
+    } xlv_targoff_u;
     STRLEN	xlv_targlen;
     SV*		xlv_targ;
     char	xlv_type;	/* k=keys .=pos x=substr v=vec /=join/re
@@ -520,6 +523,8 @@ struct xpvlv {
     char	xlv_flags;	/* 1 = negative offset  2 = negative len */
 };
 
+#define xlv_targoff xlv_targoff_u.xlvu_targoff
+
 /* This structure works in 3 ways - regular scalar, GV with GP, or fast
    Boyer-Moore.  */
 struct xpvgv {
@@ -1379,6 +1384,7 @@ sv_force_normal does nothing.
 #define LvTYPE(sv)	((XPVLV*)  SvANY(sv))->xlv_type
 #define LvTARG(sv)	((XPVLV*)  SvANY(sv))->xlv_targ
 #define LvTARGOFF(sv)	((XPVLV*)  SvANY(sv))->xlv_targoff
+#define LvSTARGOFF(sv)	((XPVLV*)  SvANY(sv))->xlv_targoff_u.xlvu_stargoff
 #define LvTARGLEN(sv)	((XPVLV*)  SvANY(sv))->xlv_targlen
 #define LvFLAGS(sv)	((XPVLV*)  SvANY(sv))->xlv_flags
 
diff --git a/t/op/array.t b/t/op/array.t
index 86740a1..0c7f533 100644
--- a/t/op/array.t
+++ b/t/op/array.t
@@ -6,7 +6,7 @@ BEGIN {
     require 'test.pl';
 }
 
-plan (127);
+plan (133);
 
 #
 # @foo, @bar, and @ary are also used from tie-stdarray after tie-ing them
@@ -466,5 +466,25 @@ package peen {
 	'arylen_p magic does not stop isa magic from being copied';
 }
 
+# [perl #118691]
+@plink=@plunk=();
+$plink[3] = 1;
+sub {
+    $_[0] = 2;
+    is $plink[0], 2, '@_ alias to nonexistent elem within array';
+    $_[1] = 3;
+    is $plink[1], 3, '@_ alias to nonexistent neg index within array';
+    is $_[2], undef, 'reading alias to negative index past beginning';
+    eval { $_[2] = 42 };
+    like $@, qr/Modification of non-creatable array value attempted, (?x:
+               )subscript -5/,
+         'error when setting alias to negative index past beginning';
+    is $_[3], undef, 'reading alias to -1 elem of empty array';
+    eval { $_[3] = 42 };
+    like $@, qr/Modification of non-creatable array value attempted, (?x:
+               )subscript -1/,
+         'error when setting alias to -1 elem of empty array';
+}->($plink[0], $plink[-2], $plink[-5], $plunk[-1]);
+
 
 "We're included by lib/Tie/Array/std.t so we need to return something true";
```
</details>

**Comment by p5pRT** [comment]
### From @cpansprout

Fixed in bbfdc87\.

I decided to go with the union\, as LvSTARGOFF is highly convenient\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

Fixed in bbfdc87\.

I decided to go with the union\, as LvSTARGOFF is highly convenient\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'open' to 'resolved'
